// contentful-migration/built/lib/action/content-type-taxonomy-validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentTypeClearTaxonomyValidationsAction = exports.ContentTypeAddTaxonomyValidationAction = exports.ContentTypeSetTaxonomyValidationsAction = void 0;
const action_1 = require("./action");
class ContentTypeSetTaxonomyValidationsAction extends action_1.EntityAction {
    constructor(contentTypeId, taxonomyValidations) {
        super();
        this.contentTypeId = contentTypeId;
        this.taxonomyValidations = taxonomyValidations;
    }
    getEntityId() {
        return this.contentTypeId;
    }
    getEntityType() {
        return action_1.EntityType.ContentType;
    }
    async applyTo(ct) {
        ct.setTaxonomyValidations(this.taxonomyValidations);
    }
}
exports.ContentTypeSetTaxonomyValidationsAction = ContentTypeSetTaxonomyValidationsAction;
class ContentTypeAddTaxonomyValidationAction extends action_1.EntityAction {
    constructor(contentTypeId, taxonomyId, linkType, options = {}) {
        super();
        this.contentTypeId = contentTypeId;
        this.taxonomyId = taxonomyId;
        this.linkType = linkType;
        this.options = options;
    }
    getEntityId() {
        return this.contentTypeId;
    }
    getEntityType() {
        return action_1.EntityType.ContentType;
    }
    async applyTo(ct) {
        ct.addTaxonomyValidation(this.taxonomyId, this.linkType, this.options);
    }
}
exports.ContentTypeAddTaxonomyValidationAction = ContentTypeAddTaxonomyValidationAction;
class ContentTypeClearTaxonomyValidationsAction extends action_1.EntityAction {
    constructor(contentTypeId) {
        super();
        this.contentTypeId = contentTypeId;
    }
    getEntityId() {
        return this.contentTypeId;
    }
    getEntityType() {
        return action_1.EntityType.ContentType;
    }
    async applyTo(ct) {
        ct.clearTaxonomyValidations();
    }
}
exports.ContentTypeClearTaxonomyValidationsAction = ContentTypeClearTaxonomyValidationsAction;
